src/extension/src/scripts/generate-test.ts

```diff
@@ -0,0 +1,126 @@
+import { CoreTemplateStudio } from "../coreTemplateStudio";
+import { CONSTANTS } from "../constants";
+
+let instance: CoreTemplateStudio;
+let backends: string[] = [];
+let frontends: string[] = [];
+let pagesObj: any[] = [];
+const projType = "FullStackWebApp";
+let syncAttemptNum = 0;
+let prevPromise: Promise<any> = Promise.resolve(null);
+
+const delay = (time: number) => {
+  return new Promise(function(resolve) {
+    setTimeout(resolve, time);
+  });
+};
+
+let getPagesObj = (
+  instance: CoreTemplateStudio,
+  frontend: any,
+  backend: any
+) => {
+  return instance
+    .getPages(projType, frontend, backend)
+    .then(pages => {
+      pagesObj = [];
+      pages.forEach((page: { name: any; templateId: any }) => {
+        pagesObj.push({
+          name: page.name,
+          identity: page.templateId
+        });
+      });
+      return pagesObj;
+    })
+    .catch((error: Error) => {
+      console.log(error.toString());
+    });
+};
+
+let generateProj = (
+  instance: CoreTemplateStudio,
+  backend: string,
+  frontend: string
+) => {
+  return getPagesObj(instance, frontend, backend).then(pagesObj => {
+    return instance.generate({
+      port: instance.getPort(),
+      payload: {
+        backendFramework: backend,
+        frontendFramework: frontend,
+        pages: pagesObj,
+        path: "../../../../../src/extension/src/template_test",
+        projectName: backend + "-" + frontend,
+        projectType: projType,
+        services: []
+      },
+      liveMessageHandler: value => {
+        value;
+      }
+    });
+  });
+};
+
+let attemptSync: any = (
+  instanceObj: CoreTemplateStudio,
+  syncAttemptNum: number
+) => {
+  if (syncAttemptNum >= CONSTANTS.API.MAX_SYNC_REQUEST_ATTEMPTS) {
+    CoreTemplateStudio.DestroyInstance();
+    throw new Error("too many failed sync requests");
+  }
+
+  return instanceObj
+    .sync({
+      port: instance.getPort(),
+      payload: { path: CONSTANTS.API.DEVELOPMENT_PATH_TO_TEMPLATES },
+      liveMessageHandler: value => {
+        value;
+      }
+    })
+    .then(() => {
```

Comment:
i feel like this nesting is unnecessary, we could do something like the following:
```ts
return instanceObj
.sync({
  // do stuff
})
.then(() => {
  return instance.getFrameworks(projType);
})
.then(frameworks => {
  //do stuff here.
})
.catch(// handle errors, could be syncattempt or something else, make sure to catch it properly.)
```
Feel like this is a bit more readable? What do you think?

Reply:
That's a good idea! It definitely helps make the code more readable. I'll add it in right now :) Thanks!